src/LSST/control/NetCommandDefs.cpp

```diff
@@ -73,6 +76,91 @@ bool NCmdSwitchCommandSource::action() {
     return result;
 }
 
+
+NCmdPower::Ptr NCmdPower::create(JsonPtr const& inJson_) {
+    auto cmd = Ptr(new NCmdPower(inJson_));
+    return cmd;
+}
+
+NCmdPower::NCmdPower(JsonPtr const& inJson) : NetCommand(inJson) {
+    LTRACE("NCmdPower::NCmdPower ", to_string(*inJson));
+    try {
+        int powerType = inJson->at("powerType");
+        switch (powerType) {
+        case MOTOR:
+            _powerType = MOTOR;
+            break;
+        case COMM:
+            _powerType = COMM;
+            break;
+        default:
+            throw NetCommandException(ERR_LOC, "unknown powerType=" + to_string(powerType) + " in " + inJson->dump());
+        }
+        _status = inJson->at("status");
+        LDEBUG(__func__, " ", getCommandName()," seqId=", getSeqId(), " powerType=", _powerType, " ", getPowerSystemTypeStr(_powerType), " status=", _status);
+    } catch (json::exception const& ex) {
+        string eMsg = string(__func__) + " constructor error in " + inJson->dump() + " what=" + ex.what();
+        LERROR(eMsg);
+        throw NetCommandException(ERR_LOC, eMsg); // &&& make a base class function to do this
+    }
+    ackJson["id"] = "ack";
+    ackJson["user_info"] = getCommandName() + " " + getPowerSystemTypeStr(_powerType) + to_string(_status);
+}
+
+NetCommand::Ptr NCmdPower::createNewNetCommand(JsonPtr const& inJson) {
+    return NCmdPower::create(inJson);
+}
+
+bool NCmdPower::action() {
+    auto context = Context::get();
+    bool result = context->model.getCurrentState()->cmdPower(_powerType, _status);
+    {
+        // Message that looks something like this needs to be broadcast
+        // {'powerType': 2, 'status': True, 'id': 'cmd_power', 'sequence_id': 123}
+        // {'id': 'powerSystemState', 'powerType': 2, 'status': true, 'state': 3 }
+        // {'id': 'powerSystemState', 'powerType': 2, 'status': true, 'state': 5}
+        // This is a duplicate broadcast and can probably be removed, but it may
+        // be useful as without this motor state is only broadcast on change.
+        json js = context->model.getPowerSystem()->getPowerSystemStateJson(_powerType);
+        string msg = to_string(js);
+        auto comServ = system::ComControlServer::get().lock();
+        if (comServ != nullptr) {
+            comServ->asyncWriteToAllComConn(msg);
```

Comment:
Why not to:
```
comServ->asyncWriteToAllComConn(js.dump());
```
Moreover, I would rewrite this method as:
```
bool NCmdPower::action() {
    auto const context = Context::get();
    if (auto const comServ = system::ComControlServer::get().lock(); comServ != nullptr) {
        json const js = context->model.getPowerSystem()->getPowerSystemStateJson(_powerType);        
        comServ->asyncWriteToAllComConn(js.dump());
    }
    return context->model.getCurrentState()->cmdPower(_powerType, _status);
}
```
I removed the long comment to illustrate the idea.
If you are using a compiler that is older than C++17 then you would do:
```
bool NCmdPower::action() {
    auto const context = Context::get();
    auto const comServ = system::ComControlServer::get().lock(); 
    if (comServ != nullptr) {
        json const js = context->model.getPowerSystem()->getPowerSystemStateJson(_powerType);        
        comServ->asyncWriteToAllComConn(js.dump());
    }
    return context->model.getCurrentState()->cmdPower(_powerType, _status);
}
```

Reply:
I think js.dump() may be expensive enough to not want it in the loop, as multiple connections should all be getting the same message, although I never expect very many concurrent connections. Whatever reason for having that in its own block has gone away. While compressing a lot of information into a single line can be desirable, I find it can make it more difficult to read the compiler errors.